ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala

```diff
@@ -721,27 +760,48 @@ private object OracleQueries extends Queries {
     path.elems.foldMap(_.fold(k => (".\"": Cord) ++ k :- '"', (_: _0.type) => "[0]"))
 
   // I cannot believe this function exists in 2021
+  // None if literal is too long
   // ORA-40454: path expression not a literal
-  private[this] def oraclePathEscape(readyPath: Cord): Fragment = {
-    val s = readyPath.toString
-    assert(
+  private[this] def oraclePathEscape(readyPath: Cord): Option[Fragment] = for {
+    readyPath <- Some(readyPath)
+    if readyPath.size <= literalStringSizeLimit
+    s = readyPath.toString
+    _ = assert(
       !s.startsWith("'") && !s.endsWith("'"),
       "Oracle JSON query syntax doesn't allow ' at beginning or ending",
     )
-    Fragment const0 ("'" + s.replace("'", "''") + "'")
-  }
+    escaped = s.replace("'", "''")
+    if escaped.length <= literalStringSizeLimit
+  } yield Fragment const0 ("'" + escaped + "'")
+  // ORA-01704: string literal too long
+  private[this] val literalStringSizeLimit = 4000
+
+  private[this] def oracleShortPathEscape(readyPath: Cord): Fragment =
+    oraclePathEscape(readyPath).getOrElse(
+      throw new IllegalArgumentException(s"path too long: $readyPath")
+    )
 
   private[http] override def equalAtContractPath(path: JsonPath, literal: JsValue): Fragment = {
     val opath: Cord = '$' -: pathSteps(path)
-    literal match {
-      case JsBoolean(_) | JsNull | JsNumber(_) | JsString(_) =>
-        val pred = opath ++ "?(@ == " ++ literal.compactPrint :- ')'
-        sql"JSON_EXISTS(" ++ contractColumnName ++ sql", " ++ oraclePathEscape(pred) ++ sql")"
-      case JsObject(_) | JsArray(_) =>
-        sql"JSON_EQUAL(JSON_QUERY(" ++ contractColumnName ++ sql", " ++ oraclePathEscape(
-          opath
-        ) ++ sql"), ${literal.compactPrint: String})"
+    def existsForm[Lit: Put](literal: Lit) =
+      (
+        "?(@ == $X)", // not a Scala interpolation
+        sql" PASSING $literal AS X",
```

Comment:
What does this change do? Does it sidestep the length limit by moving the literal out of the path? If so, let’s please add a comment.

Reply:
baa61c5fee2